### PR TITLE
Add Vector object that is templated over the type

### PR DESCRIPTION
**Pull request by garth-wells**
This is a start towards supporting multiple types.

It is very likely to evolve further, but merging now will allow other simplifications and it doesn't affect the user interface.

### CODE REVIEW

File: cpp/dolfinx/fem/DirichletBC.h [hunk 1]
```diff
@@ -203,9 +204,8 @@ class DirichletBC
   void dof_values(Eigen::Ref<Eigen::Matrix<T, Eigen::Dynamic, 1>> values) const
   {
     assert(_g);
-    la::VecReadWrapper g(_g->vector().vec());
     for (Eigen::Index i = 0; i < _dofs.rows(); ++i)
-      values[_dofs(i, 0)] = g.x[_dofs(i, 1)];
+      values[_dofs(i, 0)] = _g->x()->array()[_dofs(i, 1)];
```
Comment (chrisrichardson) [inline cpp/dolfinx/fem/DirichletBC.h:208]: Maybe these should be pulled out as const references before the loop?

Comment (garth-wells) [reply cpp/dolfinx/fem/DirichletBC.h:208]: Yes - change was made quickly.

File: cpp/dolfinx/function/Function.h [hunk 2]
```diff
@@ -74,13 +73,15 @@ class Function
   /// @return The function space
   std::shared_ptr<const FunctionSpace> function_space() const;
 
-  /// Return vector of expansion coefficients (non-const version)
+  /// Return vector of expansion coefficients at as PETSc Vec
```
Comment (chrisrichardson) [inline cpp/dolfinx/function/Function.h:76]: typo here

File: cpp/dolfinx/function/Function.cpp [hunk 3]
```diff
@@ -101,49 +99,57 @@ Function Function::collapse() const
...
+  // Check that data type is the same as the PETSc build
+  // if constexpr (std::is_same<T, PetscScalar>::value)
+  // {
+  if (!_vector)
+  {
+    _vector = create_ghosted_vector(*_function_space->dofmap()->index_map,
+                                    _x->array());
+  }
   return _vector;
+  // }
```
Comment (michalhabera) [inline cpp/dolfinx/function/Function.cpp:145]: Leftover commented code.

File: cpp/dolfinx/function/Function.h [hunk 4]
```diff
@@ -147,7 +148,9 @@ class Function
   std::shared_ptr<const FunctionSpace> _function_space;
 
   // The vector of expansion coefficients (local)
-  la::PETScVector _vector;
+  std::shared_ptr<la::Vector<PetscScalar>> _x;
+
+  // The vector of expansion coefficients
+  mutable Vec _vector = nullptr;
```
Comment (michalhabera) [inline cpp/dolfinx/function/Function.h:154]: Vector here being stored twice is weird. Is `_vector` only for caching? Is this duplicating the storage?

File: python/dolfinx/wrappers/la.cpp [hunk 5]
```diff
@@ -81,6 +82,12 @@ void la(py::module& m)
         return self[i]->vec();
       });
 
+  // dolfinx::la::VectorSpaceBasis
```
Comment (michalhabera) [inline python/dolfinx/wrappers/la.cpp:85]: VectorSpaceBasis?

File: cpp/dolfinx/function/Function.cpp (hunk 3 above)
Comment (garth-wells) [reply cpp/dolfinx/function/Function.cpp:145]: It's deliberate - will be activated immediately after OR.

File: cpp/dolfinx/function/Function.h (hunk 4 above)
Comment (garth-wells) [reply cpp/dolfinx/function/Function.h:154]: `Vec` is a PETSc vector that wraps the data in `_x`. There is no duplication of storage. `_vector` will probably move at some point once the design becomes clearer, but for now this avoids user-level changes.

**Pull request merged**